pmm/kubernetes-cluster-staging.groovy

```diff
@@ -0,0 +1,372 @@
+import hudson.model.Node.Mode
+import hudson.slaves.*
+import jenkins.model.Jenkins
+import hudson.plugins.sshslaves.SSHLauncher
+
+pipeline {
+    agent {
+        label 'awscli'
+    }
+    parameters {
+        string(
+            defaultValue: '',
+            description: 'public ssh key for "ec2-user" user, please set if you need ssh access',
+            name: 'SSH_KEY')
+        choice(
+            choices: '1\n0\n1\n2\n3\n4\n5\n6\n7\n8\n9\n10\n11\n12\n13\n14\n15\n16\n17\n18\n19\n20\n21\n22\n23\n24\n25\n26\n27\n28\n29\n30',
+            description: 'Stop the instance after, days ("0" value disables autostop and recreates instance in case of AWS failure)',
+            name: 'DAYS')
+        string(
+            defaultValue: 'true',
+            description: 'Enable Slack notification (option for high level pipelines)',
+            name: 'NOTIFY')
+        string(
+            defaultValue: 'master',
+            description: 'Tag/Branch for pmm-qa repository',
+            name: 'PMM_QA_GIT_BRANCH')
+        string(
+            defaultValue: '',
+            description: 'Commit hash for pmm-qa branch',
+            name: 'PMM_QA_GIT_COMMIT_HASH')
+    }
+    options {
+        skipDefaultCheckout()
+    }
+
+    stages {
+        stage('Prepare') {
+            steps {
+                deleteDir()
+                wrap([$class: 'BuildUser']) {
+                    sh """
+                        echo "\${BUILD_USER_EMAIL}" > OWNER_EMAIL
+                        echo "\${BUILD_USER_EMAIL}" | awk -F '@' '{print \$1}' > OWNER_FULL
+                        echo "pmm-kubernetes-cluster-\$(cat OWNER_FULL)-\$(date -u '+%Y%m%d%H%M%S')-${BUILD_NUMBER}" \
+                            > VM_NAME
+                    """
+                }
+                script {
+                    def OWNER = sh(returnStdout: true, script: "cat OWNER_FULL").trim()
+                    def OWNER_EMAIL = sh(returnStdout: true, script: "cat OWNER_EMAIL").trim()
+                    def OWNER_SLACK = slackUserIdFromEmail(botUser: true, email: "${OWNER_EMAIL}", tokenCredentialId: 'JenkinsCI-SlackBot-v2')
+
+                    if ("${NOTIFY}" == "true") {
+                        slackSend botUser: true, channel: '#pmm-ci', color: '#FFFF00', message: "[${JOB_NAME}]: build started - ${BUILD_URL}"
+                        slackSend botUser: true, channel: "@${OWNER_SLACK}", color: '#FFFF00', message: "[${JOB_NAME}]: build started - ${BUILD_URL}"
+                    }
+                }
+            }
+        }
+
+        stage('Run VM') {
+            steps {
+                withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'AMI/OVF', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
+                    sh '''
+                        export VM_NAME=\$(cat VM_NAME)
```

Comment:
The lines of code 65-92 are the exact same in several pipelines, so they better be a function to avoid repeating the same multiple times.

Reply:
Do you mean like a shared library?

Reply:
Not really. We already have such functions in https://github.com/Percona-Lab/jenkins-pipelines/tree/master/vars.

Reply:
I have addressed this: https://github.com/Percona-Lab/jenkins-pipelines/pull/744 merged this PR only because I could not test without merging to a branch.